Library/Homebrew/keg_relocate.rb

```diff
@@ -165,6 +165,57 @@ def replace_text_in_files(relocation, files: nil)
     changed_files
   end
 
+  def relocate_build_prefix(keg, old_prefix, new_prefix, verbose)
+    # Find binaries which match prefix strings.
+    string_matches = Set.new
+    keg.each_unique_file_matching(old_prefix) do |file|
+      string_matches << file
+    end
+
+    # Remove files that are known to be plain text.
+    # Note that this also removes sharballs, which we currently don't patch
+    # because they appear to break after patching.
+    string_matches -= text_files
+
+    # Only keep files with prefix matches which contain null bytes.  We need to do this
+    # because text_files is not intended to identify all types of text files.
+    binary_string_matches = Set.new
+    keg.each_unique_binary_file(string_matches) do |file|
+      binary_string_matches << file
+    end
+
+    # Split binary by null characters into array and substitute new cellar for old cellar.
+    # Null padding is added if the new string is too short.
+    binary_string_matches.each do |binary_file|
+      binary_file.ensure_writable do
+        binary = File.binread binary_file
+        puts "Replacing build prefix in: #{binary_file}" if verbose
+        binary_strings = binary.split(NULL_BYTE)
+        match_indices = binary_strings.each_index.select { |i| binary_strings[i].include?(old_prefix) }
+
+        # Only perform substitution on strings which match prefix regex.
+        match_indices.each do |i|
+          s = binary_strings[i]
+          binary_strings[i] = s.gsub(old_prefix, new_prefix)
+                               .ljust(s.size, NULL_BYTE)
+        end
+
+        # Add back null padding at the end of the binary if needed.
+        patched_binary = binary_strings.join(NULL_BYTE).ljust(binary.size, NULL_BYTE)
```

Comment:
The reason I had to do this was because by default, `.split` truncates trailing empty strings.  So if the last few bytes of the binary are null bytes (i.e. `some_string\x00\x00\x00\x00`, then the last strings in `binary_strings` will be empty and be dropped.  However, I see here that if I were to set the optional `limit` parameter of `.split` to `-1`, the empty strings are not dropped:  https://www.geeksforgeeks.org/ruby-string-split-method-with-examples/.  I was not aware of this before but I think it's worth trying to implement, because if it works we could get rid the second `ljust` entirely, making this much simpler.

Reply:
Can we rule out leading null bytes? If not, then we may want to split on a regex instead of just a raw null byte. See https://ruby-doc.org/core-2.6.3/String.html#method-i-split

Reply:
I believe leading null bytes don't normally occur in ELF/MachO binaries, but this may not be true in other binary file types so I don't know that we can rule them out.  From my current testing, it seems that using a regex works just as well and may be safer, though I still have to add a limit of -1 to avoid dropping the trailing empty strings.